File: frontend/src/service/api/cron.ts
import { request } from '../request';

export type CronScriptMode = 'inline' | 'file';
export type CronTriggerMode = 'manual' | 'schedule' | string;

export interface CronTaskItem {
  id: number;
  name: string;
  schedule: string;
  script: string;
  scriptMode: CronScriptMode;
  currentFileId: number;
  currentFileName: string;
  currentFileVersion: number;
  currentFilePath: string;
  currentFileSha256: string;
  status: number;
  timeout: number;
  nextRun: string;
  createdAt: string;
  updatedAt: string;
}

export interface CronTaskFileItem {
  id: number;
  taskId: number;
  version: number;
  originalName: string;
  storedName: string;
  filePath: string;
  sizeBytes: number;
  sha256: string;
  isCurrent: boolean;
  createdAt: string;
}

export interface CronLogItem {
  id: number;
  taskId: number;
  taskName: string;
  startTime: string;
  endTime: string;
  status: number;
  exitCode: number;
  output: string;
  error: string;
  duration: number;
  triggerMode: CronTriggerMode;
  scriptMode: CronScriptMode;
  scriptSnapshot: string;
  scriptFileId: number;
  scriptFileVersion: number;
  scriptFileName: string;
  scriptFilePath: string;
  scriptFileSha256: string;
}

export interface CronTaskListResp {
  list: CronTaskItem[];
  total: number;
}

export interface CronTaskFileListResp {
  list: CronTaskFileItem[];
  total: number;
}

export interface CronLogListResp {
  list: CronLogItem[];
  total: number;
}

export interface CronTaskFormPayload {
  name: string;
  schedule: string;
  scriptMode?: CronScriptMode;
  script?: string;
  status: number;
  timeout: number;
}

export interface CronTaskUpdatePayload extends Partial<CronTaskFormPayload> {
  status?: number;
  timeout?: number;
}

export function fetchCronTaskList(params?: { page?: number; pageSize?: number; name?: string }) {
  return request<CronTaskListResp>({
    url: '/api/cron/task',
    method: 'get',
    params
  });
}

export function createCronTask(data: CronTaskFormPayload) {
  return request<void>({
    url: '/api/cron/task',
    method: 'post',
    data
  });
}

export function updateCronTask(id: number, data: CronTaskUpdatePayload) {
  return request<void>({
    url: `/api/cron/task/${id}`,
    method: 'put',
    data
  });
}

export function deleteCronTask(id: number) {
  return request<void>({
    url: `/api/cron/task/${id}`,
    method: 'delete'
  });
}

export function triggerCronTask(id: number) {
  return request<void>({
    url: `/api/cron/task/${id}/trigger`,
    method: 'post'
  });
}

export function fetchCronTaskScriptHistory(taskId: number, params?: { page?: number; pageSize?: number }) {
  return request<CronTaskFileListResp>({
    url: `/api/cron/task/${taskId}/script/history`,
    method: 'get',
    params
  });
}

export function uploadCronTaskScript(taskId: number, data: FormData) {
  return request<void>({
    url: `/api/cron/task/${taskId}/script`,
    method: 'post',
    data
  });
}

export function activateCronTaskScript(taskId: number, fileId: number) {
  return request<void>({
    url: `/api/cron/task/${taskId}/script/${fileId}/activate`,
    method: 'post'
  });
}

export function fetchCronLogList(params?: { page?: number; pageSize?: number; taskId?: number; status?: number }) {
  return request<CronLogListResp>({
    url: '/api/cron/log',
    method: 'get',
    params
  });
}

export function fetchCronLogDetail(id: number) {
  return request<CronLogItem>({
    url: `/api/cron/log/${id}`,
    method: 'get'
  });
}

File: backend/common/cron/cron.go
package cron

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"logflux/internal/config"
	"logflux/internal/types"
	cronmodel "logflux/model/cron"
)

const (
	ScriptModeInline = "inline"
	ScriptModeFile   = "file"

	TriggerModeManual   = "manual"
	TriggerModeSchedule = "schedule"

	DefaultFilesDir = "/config/cron-files"
)

type uploadContextKey string

const (
	uploadTempPathKey uploadContextKey = "cron_upload_temp_path"
	uploadFileNameKey uploadContextKey = "cron_upload_file_name"
)

func FilesBaseDir(c *config.Config) string {
	if c != nil {
		if dir := strings.TrimSpace(c.CronFilesDir); dir != "" {
			return filepath.Clean(dir)
		}
	}
	return DefaultFilesDir
}

func EnsureWorkspace(baseDir string) error {
	baseDir = filepath.Clean(strings.TrimSpace(baseDir))
	if baseDir == "" {
		return fmt.Errorf("定时任务脚本目录为空")
	}
	for _, dir := range []string{
		baseDir,
		filepath.Join(baseDir, "tasks"),
		filepath.Join(baseDir, "staging"),
	} {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return fmt.Errorf("创建目录失败: %s, %w", dir, err)
		}
	}
	return nil
}

func TaskDir(baseDir string, taskID uint) string {
	return filepath.Join(filepath.Clean(strings.TrimSpace(baseDir)), "tasks", fmt.Sprintf("%d", taskID))
}

func TempDir(baseDir string) string {
	return filepath.Join(filepath.Clean(strings.TrimSpace(baseDir)), "staging")
}

func SafeFileName(name string) string {
	base := strings.TrimSpace(name)
	base = filepath.Base(base)
	base = strings.ReplaceAll(base, "/", "_")
	base = strings.ReplaceAll(base, "\\", "_")
	base = strings.ReplaceAll(base, "..", "_")
	base = strings.TrimSpace(base)
	if base == "" {
		return "script.sh"
	}
	return base
}

func NormalizeScriptMode(mode string) string {
	mode = strings.ToLower(strings.TrimSpace(mode))
	switch mode {
	case "", ScriptModeInline:
		return ScriptModeInline
	case ScriptModeFile:
		return ScriptModeFile
	default:
		return mode
	}
}

func ValidateScriptMode(mode string) error {
	switch NormalizeScriptMode(mode) {
	case ScriptModeInline, ScriptModeFile:
		return nil
	default:
		return fmt.Errorf("脚本来源类型无效")
	}
}

func ScriptModeLabel(mode string) string {
	switch NormalizeScriptMode(mode) {
	case ScriptModeFile:
		return "文件脚本"
	default:
		return "内联脚本"
	}
}

func DefaultScriptText(script string) string {
	return strings.TrimSpace(script)
}

func ReadTaskScriptMode(task *cronmodel.CronTask) string {
	if task == nil {
		return ScriptModeInline
	}
	return NormalizeScriptMode(task.ScriptMode)
}

func WithUploadTempPath(ctx context.Context, tempPath string) context.Context {
	return context.WithValue(ctx, uploadTempPathKey, tempPath)
}

func WithUploadFileName(ctx context.Context, fileName string) context.Context {
	return context.WithValue(ctx, uploadFileNameKey, fileName)
}

func UploadTempPathFromContext(ctx context.Context) string {
	if ctx == nil {
		return ""
	}
	value, _ := ctx.Value(uploadTempPathKey).(string)
	return strings.TrimSpace(value)
}

func UploadFileNameFromContext(ctx context.Context) string {
	if ctx == nil {
		return ""
	}
	value, _ := ctx.Value(uploadFileNameKey).(string)
	return strings.TrimSpace(value)
}

func NormalizeTriggerMode(triggerMode string) string {
	triggerMode = strings.ToLower(strings.TrimSpace(triggerMode))
	switch triggerMode {
	case TriggerModeSchedule, TriggerModeManual:
		return triggerMode
	case "":
		return TriggerModeManual
	default:
		return triggerMode
	}
}

func ExecutionTimeout(timeout int) time.Duration {
	if timeout <= 0 {
		timeout = 60
	}
	return time.Duration(timeout) * time.Second
}

func IsPathWithinBase(baseDir, targetPath string) bool {
	baseDir = filepath.Clean(strings.TrimSpace(baseDir))
	targetPath = filepath.Clean(strings.TrimSpace(targetPath))
	if baseDir == "" || targetPath == "" {
		return false
	}

	rel, err := filepath.Rel(baseDir, targetPath)
	if err != nil {
		return false
	}
	rel = filepath.Clean(rel)
	return rel != ".." && !strings.HasPrefix(rel, ".."+string(filepath.Separator))
}

func CronFileItem(file *cronmodel.CronTaskFile) types.CronTaskFileItem {
	if file == nil {
		return types.CronTaskFileItem{}
	}
	return types.CronTaskFileItem{
		ID:           file.ID,
		TaskID:       file.TaskID,
		Version:      file.Version,
		OriginalName: file.OriginalName,
		StoredName:   file.StoredName,
		FilePath:     file.FilePath,
		SizeBytes:    file.SizeBytes,
		SHA256:       file.SHA256,
		IsCurrent:    file.IsCurrent,
		CreatedAt:    file.CreatedAt.Format("2006-01-02 15:04:05"),
	}
}

func CronTaskItem(task *cronmodel.CronTask, currentFile *cronmodel.CronTaskFile, nextRun string) types.CronTaskItem {
	if task == nil {
		return types.CronTaskItem{}
	}
	item := types.CronTaskItem{
		ID:            task.ID,
		Name:          task.Name,
		Schedule:      task.Schedule,
		Script:        task.Script,
		ScriptMode:    NormalizeScriptMode(task.ScriptMode),
		CurrentFileID: task.CurrentFileID,
		Status:        task.Status,
		Timeout:       task.Timeout,
		NextRun:       nextRun,
		CreatedAt:     task.CreatedAt.Format("2006-01-02 15:04:05"),
		UpdatedAt:     task.UpdatedAt.Format("2006-01-02 15:04:05"),
	}
	if currentFile != nil {
		item.CurrentFileID = currentFile.ID
		item.CurrentFileName = currentFile.OriginalName
		item.CurrentFileVersion = currentFile.Version
		item.CurrentFilePath = currentFile.FilePath
		item.CurrentFileSHA256 = currentFile.SHA256
	}
	return item
}

func CronLogItem(logEntry *cronmodel.CronTaskLog) types.CronLogItem {
	if logEntry == nil {
		return types.CronLogItem{}
	}
	endTime := ""
	if !logEntry.EndTime.IsZero() {
		endTime = logEntry.EndTime.Format("2006-01-02 15:04:05")
	}
	return types.CronLogItem{
		ID:                logEntry.ID,
		TaskID:            logEntry.TaskID,
		TaskName:          logEntry.TaskName,
		StartTime:         logEntry.StartTime.Format("2006-01-02 15:04:05"),
		EndTime:           endTime,
		Status:            logEntry.Status,
		ExitCode:          logEntry.ExitCode,
		Output:            logEntry.Output,
		Error:             logEntry.Error,
		Duration:          logEntry.Duration,
		TriggerMode:       logEntry.TriggerMode,
		ScriptMode:        NormalizeScriptMode(logEntry.ScriptMode),
		ScriptSnapshot:    logEntry.ScriptSnapshot,
		ScriptFileID:      logEntry.ScriptFileID,
		ScriptFileVersion: logEntry.ScriptFileVersion,
		ScriptFileName:    logEntry.ScriptFileName,
		ScriptFilePath:    logEntry.ScriptFilePath,
		ScriptFileSHA256:  logEntry.ScriptFileSHA256,
	}
}
